### PR TITLE
MAID 2015 Background res proof

### PR DESCRIPTION
**Pull request by dhardy**
Note: builds will fail immediately until resource_proof 0.4.0 is published.
https://github.com/maidsafe/resource_proof/pull/3

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @afck

(maidsafe_highfive has picked a reviewer for you, use r? to override)

**Comment by dhardy** [comment]
So, checks passed and Fraser's comments addressed, except the one about unit tests for resource proof (no, I wasn't proposing to do that within this PR).

File: src/node.rs [hunk 1]
```diff
@@ -113,8 +113,8 @@ impl NodeBuilder {
                           -> (RoutingActionSender, StateMachine) {
         let full_id = FullId::new();
 
-        StateMachine::new(move |crust_service, timer, outbox2| if self.first {
-                              if let Some(state) = states::Node::first(self.cache,
+        StateMachine::new(move |action_sender, crust_service, timer, outbox2| if self.first {
+                              if let Some(state) = states::Node::first(action_sender, self.cache,
```
Comment (afck) [inline src/node.rs]: (That looks like a rustfmt bug…)

File: src/resource_prover.rs [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::time::{Duration, Instant};
+use timer::Timer;
+use types::RoutingActionSender;
+use utils::DisplayDuration;
+
+/// Time (in seconds) between accepting a new candidate (i.e. receiving an `AcceptAsCandidate` from
+/// our section) and sending a `CandidateApproval` for this candidate.  If the candidate cannot
```
Comment (afck) [inline src/resource_prover.rs]: Only one space after full stops. (Also below.)

File: src/resource_prover.rs [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+        if self.response_parts.is_empty() {
+            // `GetNodeNameResponse` has timed out.
+            info!("{} Failed to get relocated name from the network, so restarting.",
+                  log_ident);
+            outbox.send_event(Event::RestartRequired);
+        } else {
+            // `NodeApproval` has timed out.
+            let completed = self.response_parts
+                .values()
+                .filter(|parts| parts.is_empty())
```
Comment (afck) [inline src/resource_prover.rs]: `filter(Vec::is_empty)`?

Comment (dhardy) [inline src/resource_prover.rs]: Type error (requires double reference to a `Vec`).

**Pull request merged**